Mathlib/Order/BourbakiWitt.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2025 Finn Mortimore. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Finn Mortimore
+-/
+import Mathlib.Order.Preorder.Chain
+import Mathlib.Data.Set.Lattice
+import Mathlib.Dynamics.FixedPoints.Basic
+
+/-!
+# Bourbaki-Witt Theorem
+
+This file proves the Bourbaki-Witt Theorem.
+
+## Main definitions
+
+- class `ChainCompletePartialOrder` : A non-empty partial order is a chain complete partial order
+  such that every non-empty chain has a supremum
+- structure `Inflationary` : Let $X$ be a partial order. A function $f : X → X$ is inflationary if,
+  for all $x ∈ X$, $x ≤ f x$
+
+## Main statements
+
+- `bourbaki_witt` : The Bourbaki-Witt Theorem : If $X$ is a chain complete
+partial order and $f : X → X$ is inflationary, then $f$ has a fixed point
+
+## References
+
+The proof used can be found in [serge_lang_algebra]
+-/
+
+/-- The type of non-empty chains of an order -/
+@[ext]
+structure NonemptyChain (α : Type) [LE α] where
+  /-- The underlying set of a non-empty chain -/
+  carrier : Set α
+  /-- By definition, a non-empty chain is non-empty -/
+  Nonempty' : carrier.Nonempty
+  /-- By definition, a non-empty chain is a chain -/
+  isChain' : IsChain (· ≤ ·) carrier
+
+instance {α : Type} [LE α] : SetLike (NonemptyChain α) α where
+  coe := NonemptyChain.carrier
+  coe_injective' _ _ := NonemptyChain.ext
+
+/-- A chain complete partial order is a non-empty partial order such that every chain has a supremum
+(which we call `cSup`) -/
+class ChainCompletePartialOrder (α : Type) [Nonempty α] extends PartialOrder α where
```

Comment:
We already have a definition in Mathlib of `OmegaCompletePartialOrder`. There, ⍵-chains are expressed as monotone functions `ℕ →o α`. Did you decide here to use a subset which is totally ordered because that is what `IsChain` uses? (My impression is that `IsChain` is indeed nicer to work with here.)

Could you add an instance that `ChainCompletePartialOrder` implies `OmegaCompletePartialOrder`? And similarly that `CompleteLattice` implies `ChainCompletePartialOrder`?

We should also add [the equivalence with DCPOs](https://en.wikipedia.org/wiki/Complete_partial_order#Characterizations), but that could also be follow-up work, doesn't need to be part of this PR.

Reply:
We tend to leave out assumptions like `Nonempty` when defining new structures, and add these as assumptions only when they are really needed.
```suggestion
class ChainCompletePartialOrder (α : Type) extends PartialOrder α where
```